[Ingest] Remove meta processor and expose templates in other processors

martijnvg opened the pull request:
* Added template infrastructure via FieldPathTemplate (adds template support for field like settings) and ValueSource (adds template support for value like settings.
* Adjusted the `set` and `remove` processor to be able to use templates via the new infrastructure. The `field` and `value` settings are now templateble.
* Removed the `meta` processor, since any processor can now access and modify meta fields.

PR for #14990

rmuir commented:
why can't this plugin use mustache via the scriptengine service? Sorry: I don't think we should have secret embedded script engines inside other plugins. I have found this before and it seriously pisses me off, we can't do this.

martijnvg commented:
@rmuir That is the plan. The ingest framework has been directly using mustache, because it can't directly depend on the script service. It needs to be ES agnostic. In a follow up PR, I plan to add a `TemplateService` (or something like that) that the ingest framework will use directly. The implementation in the ingest plugin will then use `lang-mustache ` module via the script service.

rmuir commented:
> The ingest framework has been directly using mustache, because it can't directly depend on the script service. It needs to be ES agnostic

This makes no sense. 


> In a follow up PR, I plan to add a TemplateService (or something like that) that the ingest framework will use directly.

This makes no sense.

Please, lets just use the scriptengineservice, and lookup "mustache", it is really just that simple.

uboness commented:
> This makes no sense.

why? The processors themselves cannot depend on ES specific codebase. The reason for this is that we want to reuse the processors in LS. So in order to have processors use templates, we need to have some sort of a Template abstraction (I think that's what @martijnvg refers to with `TemplateService`) that is part of the ingest code (at least the code that will be shared with LS).

rmuir commented:
This does not justify forking scripting engines, sorry. It needs to go thru the script engine service.

nik9000 commented:
> The processors themselves cannot depend on ES specific codebase.

I wonder if that is a good excuse to yank the some of the ES code base into a library you can share.

rmuir commented:
our script engine support is not something that needs to become a library, its out of the question. We do everything possible to contain their craziness.

martijnvg commented:
@rmuir @uboness I've updated the PR. Ingest no longer has a mustache compile time dependency and indirectly depends on the lang-mustache module. I've added a `TemplateService` interface to avoid that the ingest framework gets a compile dependency on ES. The ingest plugin provides an implementation that uses ES' script service.

> I wonder if that is a good excuse to yank the some of the ES code base into a library you can share.

@nik9000 Indirectly using the script service in ingest avoids granting permissions in ingest that a template engine may need. Also it has already been granted in the lang-moustache module, so no need to grant twice.

javanna commented on docs/plugins/ingest.asciidoc:
```diff
@@ -403,57 +403,110 @@ An example that adds the parsed date to the `timestamp` field based on the `init
 
-The following example changes the index of a document to `alternative_index` instead of indexing it into an index
-that was specified in the index or bulk request:
+Accessing a field in the source is straight forward and one can refer to fields by
```
straightforward is a single word ;)

javanna commented on docs/plugins/ingest.asciidoc:
```diff
@@ -403,57 +403,110 @@ An example that adds the parsed date to the `timestamp` field based on the `init
+The following metadata fields are accessible by a processor: `_index`, `_type`, `_id`, `_routing`, `_parent`,
+`_timestamp` and `_ttl`.
+
+Beyond metadata fields and source fields, the ingest plugin also adds ingest metadata to documents being processed.
```
shall we clarify that these fields are transient and won't be indexed?

javanna commented on docs/plugins/ingest.asciidoc:
```diff
@@ -403,57 +403,110 @@ An example that adds the parsed date to the `timestamp` field based on the `init
+
+As opposed to Elasticsearch metadata fields, the ingest metadata field name `_ingest` isn't restricted by Elasticsearch
+and is a valid field name in the source of a document, so there is a small mismatch.
+In the ingest plugin when is referred to `_ingest`  in a field setting or template then ingest metadata is accessed.
```
Looking to simplify this block: As opposed to Elasticsearch metadata fields, the ingest metadata field name `_ingest` can be used as a valid field name in the source of a document. Use `_source._ingest` to refer to it, otherwise `_ingest` will be interpreted as ingest metadata fields by the ingest plugin.

javanna commented on plugins/ingest/src/main/java/org/elasticsearch/ingest/processor/geoip/GeoIpProcessor.java:
```diff
@@ -70,26 +71,23 @@ public void execute(IngestDocument ingestDocument) {
+                default:
+                    throw new IllegalStateException("Unsupported database type [" + dbReader.getMetadata().getDatabaseType() + "]");
+            }
+            // since we already create a map instance per document we supply a lamda that directly returns the geoData map.
```
s/lamda/lambda

javanna commented on plugins/ingest/src/main/java/org/elasticsearch/ingest/processor/split/SplitProcessor.java:
```diff
@@ -58,7 +58,8 @@ public void execute(IngestDocument document) {
             throw new IllegalArgumentException("field [" + field + "] is null, cannot split.");
         }
-        document.setFieldValue(field, Arrays.asList(oldVal.split(separator)));
+        // since we already create a list instance per document we supply a lamda that directly returns the list.
```
s/lamda/lambda :)

javanna commented on plugins/ingest/src/main/java/org/elasticsearch/ingest/IngestDocument.java:
```diff
@@ -155,6 +155,17 @@ public boolean hasField(String path) {
      * @throws IllegalArgumentException if the path is null, empty, invalid or if the field doesn't exist.
      */
     public void removeField(String path) {
+        removeField(TemplateService.wrap(path));
```
I wonder what the final goal is here. Do we want to keep both variations 1) string argument 2) template argument or will the string one go away?

martijnvg replied:
I think that depends on if all or almost all processors end up using templates. Right now in this PR only remove and set processors use templates, so I kept the old method signatures around to not make the change bigger than it is.

javanna commented on plugins/ingest/src/main/java/org/elasticsearch/ingest/IngestDocument.java:
```diff
@@ -231,7 +242,7 @@ private static Object resolve(String pathElement, String fullPath, Object contex
      */
     public void appendFieldValue(String path, Object value) {
-        setFieldValue(path, value, true);
+        setFieldValue(TemplateService.wrap(path), ValueSource.wrap(value), true);
```
does this mean that a path provided as string doesn't support a template? I guess so and that way the processors will choose whether they support templating or not rather than supporting it at the IngestDocument level? I wonder if we need this wrapping though for simple strings. Given that a template in the end returns a string, do we really need to wrap a string into the StringTemplate? seems like strings could just skip the execution part as a whole?

martijnvg replied:
> does this mean that a path provided as string doesn't support a template?

No. If processors impl like to support templates they can can use the method variant that does support templating. I the case of this method, we just need to add `public void appendFieldValue(Template path, ValueSource value) {...}` because no processor is using it yet. Or better just change to current method to accept Template and ValueSource, since I think the `push` processor will also support templating.

javanna replied:
ok but I think the templated methods should call the simple ones providing the result of the templates execution and not the other way around?

martijnvg replied:
I see what you mean, that makes sense.

javanna commented on plugins/ingest/src/main/java/org/elasticsearch/ingest/IngestDocument.java:
```diff
@@ -332,6 +355,16 @@ private void setFieldValue(String path, Object value, boolean append) {
+        // If there is a field in the source with the name '_ingest' it gets overwritten here,
+        // if access to that field is required then it get accessed via '_source._ingest'
+        model.put("_ingest", ingestMetadata);
+        return model;
```
looking for a way to have the same logic that we already have in IngestDocument here too... the behaviour needs to be unified either via templates or via getFieldValue. I am not sure that splitting source and metadata again is the solution, cause it will complicate things again on the getFieldValue/setFieldValue side of things. need to think about this some more.

javanna replied:
I think the goal here is to have exactly the same behaviour that we have in IngestDocument (FieldPath). The way IngestDocument works is that _id and _source._id are equivalent, so I think everything is fine. the _source prefix won't be used in practice that often, it's there only to make sure that _source._ingest can be accessed, and one cannot really have _id etc. as part of the source.
The only thing that we could improve here is that if we change logic in FieldPath, we have to remember to adapt what is exposed to templates too, which we will most likely forget about :) I don't have a good solution for this yet.

martijnvg replied:
I see, not a big deal. So this can remain this way.

javanna commented on plugins/ingest/src/main/java/org/elasticsearch/ingest/ValueSource.java:
```diff
@@ -0,0 +1,196 @@
+            return new SimpleValue(value);
+        } else if (value instanceof String) {
+            if (templateService == null) {
+                return new SimpleValue(value);
```
are we sure we want to silently go ahead this way when templateService is null?

martijnvg replied:
this is executed in the case we just wrap a value. (in the ingest methods that just accept a Object as value). I can remove it and instead add a noop template service impl and use that in the cases where just a value is being wrapped?

javanna replied:
I think this may relate with another comment I left. I don't follow why string/object values that are not templates need to become templates that do nothing. To me, given that templates simply allow to modify which field and/or which value to get/set, they should be an additional step that happens before the usual get/set/remove that accepts a string and/or an object. Makes sense?

martijnvg replied:
right, so field the `path` parameters in `IngestDocument` should just be strings?

javanna replied:
this other comment summarizes what I mean, hopefully :) https://github.com/elastic/elasticsearch/pull/15415/files#r47890764

javanna commented on plugins/ingest/src/main/java/org/elasticsearch/ingest/ValueSource.java:
```diff
@@ -0,0 +1,196 @@
+        }
+    }
+
+    final class SimpleValue implements ValueSource {
```
call it ObjectValue ?

javanna commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java:
```diff
@@ -266,8 +275,41 @@ private Client client() {
+        @Override
+        public Template innerCompile(String template) {
+            Script script = new Script(template, ScriptService.ScriptType.INLINE, "mustache", Collections.emptyMap());
+            CompiledScript compiledScript = scriptService.compile(script, INSTANCE, null /* only needed for indexed scripts */);
```
not sure I get the comment, do we make use of indexed templates in ingest?

martijnvg replied:
no, it means the null can be specified because ingest isn't using indexed scripts. I'll update the comment :)

javanna replied:
ok thanks

javanna commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java:
```diff
@@ -266,8 +275,41 @@ private Client client() {
 
+        @Override
+        public Template innerCompile(String template) {
+            Script script = new Script(template, ScriptService.ScriptType.INLINE, "mustache", Collections.emptyMap());
```
you can use MustacheScriptEngineService.NAME

martijnvg replied:
no :) because `MustacheScriptEngineService` class is part of the lang-mustache module and we can't depend on it. Maybe we should add a constant in the `InternalTemplateService`?

javanna replied:
so we depend on es core which provides us with mustache which we rely on but we can't depend on it directly? but lang-mustache will be there all the time?

martijnvg replied:
> so we depend on es core which provides us with mustache

No, core doesn't provide it. It merely provides the interfaces script engines can implement. 

> but lang-mustache will be there all the time?

Yes, it is a plugin that is installed by default. So when the ingest plugin starts up we can use it by specifying that we want to use `mustache` language.

javanna replied:
ok thanks for the explanation.

javanna commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java:
```diff
@@ -266,8 +275,41 @@ private Client client() {
+
+    static class InternalTemplateService implements TemplateService {
+
+        public static final ScriptContext.Plugin INSTANCE = new ScriptContext.Plugin("elasticsearch-ingest", "ingest");
```
call it INGEST_SCRIPT_CONTEXT instead?

javanna commented on plugins/ingest/src/main/java/org/elasticsearch/ingest/TemplateService.java:
```diff
@@ -0,0 +1,70 @@
+public interface TemplateService {
+
+    default Template compile(String template) {
+        if (template != null) {
```
not sure about this default impl. Seems like the runtime has to plug in its own template service and template impls. but in the interface we assume that we always have mustache. I don't mean to make the engine pluggable, but I think we have to better figure out who needs to do what and how when the ingest library is used outside of the context of es core. I think for now this assumption would be better placed as part of the implementation.

martijnvg replied:
I think right now the only thing that we assume is that the template markers are `{{` and `}}`. Whether mustache or another templating engine is used is up to the runtime the pipeline framework runs in.

So with that in mind I do think it is ok to have logic that decides whether we delegate to the actual template service implementation. But maybe we should do this in a different place then here?

javanna replied:
> But maybe we should do this in a different place then here?

yes in the impl... :) sorry but I find this confusing.

martijnvg replied:
ok, let me think about this :)

javanna commented on plugins/ingest/src/main/java/org/elasticsearch/ingest/ValueSource.java:
```diff
@@ -0,0 +1,196 @@
+
+        @Override
+        public Object copy(Map<String, Object> model) {
+            return template.execute(model);
```
seems like we are mixing up the deep copy aspect with the execution of the template. would it make sense to keep them separated and expose the Template object that we already have for the field path? The two look so different but in the end the only difference is that we have to do a deep copy upfront right?

martijnvg replied:
the reason that this is here, is because a value can be templated too:
* a simple value "abc{{model.field1}}"
* or a complex value like { "field1" : "no templates", "field2" : "templates {{model.field1}}"

So in the case a value is a map, while we copy we also resolve templates. The actual template execution is delegated to the template impl, so that isn't so bad after all? Also if we did templating somewhere else we would need to evaluate each value twice...

javanna replied:
good point. Then the name of the method (copy) in the interface lies :)  but that also means we have to come up with a better name, which is never easy.  copyAndResolve ? or just resolve?

martijnvg replied:
Lets use `copyAndResolve()`? It is then clear that it then copies and resolves templates.

javanna replied:
sounds good

javanna commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java:
```diff
@@ -266,8 +275,41 @@ private Client client() {
+
+    }
+
+    static class InternalTemplateService implements TemplateService {
```
I am starting to wonder if this additional code belongs to PipelineStore. seems like it's become much more than just a store.

martijnvg replied:
Agreed, so lets move some inner classes to top level ones? Like `ProcessorFactoryProvider` and `InternalTemplateService `?

javanna replied:
+1

javanna commented on plugins/ingest/src/test/java/org/elasticsearch/ingest/IngestDocumentTests.java:
```diff
@@ -621,6 +619,34 @@ public void testRemoveEmptyField() {
+        ingestDocument.setFieldValue(templateService.compile("{{_index}}"), ValueSource.wrap("{{_index}}", templateService));
+        assertThat(ingestDocument.getFieldValue("index", String.class), equalTo("index"));
+        ingestDocument.removeField(templateService.compile("{{_index}}"));
+        assertThat(ingestDocument.hasField("index"), is(false));
```
oh boy, so this test sets a field called "index" to value "index" which is contained in the "_index" field. I must admit I would try to make it a bit easier to read using different values if possible.

martijnvg commented:
@javanna I've updated this PR.

javanna commented on docs/plugins/ingest.asciidoc:
```diff
@@ -403,57 +403,110 @@ An example that adds the parsed date to the `timestamp` field based on the `init
 --------------------------------------------------
 {
-  "description" : "...",
-  "processors" : [
```
I see that you removed the processors array here, did you do that on purpose?

martijnvg replied:
it used to be part of the `meta` processor config. I thought I could shorten the examples by only showing a processor, since the previous section shows entire pipeline examples.

javanna replied:
ok maybe it's fine, was just curious

javanna commented on plugins/ingest/src/main/java/org/elasticsearch/ingest/IngestDocument.java:
```diff
@@ -332,6 +352,15 @@ private void setFieldValue(String path, Object value, boolean append) {
+        model.put("_source", sourceAndMetadata);
+        // If there is a field in the source with the name '_ingest' it gets overwritten here,
+        // if access to that field is required then it get accessed via '_source._ingest'
+        model.put("_ingest", ingestMetadata);
```
maybe we should have a constant somewhere with _source and _ingest? pretty sure we already use them somewhere else

javanna commented on plugins/ingest/src/main/java/org/elasticsearch/ingest/ValueSource.java:
```diff
@@ -0,0 +1,190 @@
+/**
+ * Holds a value. If the value is requested a copy is made and optionally template snippets are resolved too.
+ */
+// The best name I could come up with!
```
can't come up with any better ones :) we can remove this comment then?

javanna commented on plugins/ingest/src/test/java/org/elasticsearch/ingest/TestTemplateService.java:
```diff
@@ -0,0 +1,64 @@
+
+    public static TemplateService instance() {
+        return INSTANCE;
+    }
```
maybe it would be better if each test had its own instance of TestTemplateService (better test isolation)? I think we shouldn't worry about performance or too many objects created here.

martijnvg replied:
makes sense

javanna commented on plugins/ingest/src/test/java/org/elasticsearch/ingest/TestTemplateService.java:
```diff
@@ -0,0 +1,64 @@
+        return INSTANCE;
+    }
+
+    public static Template template(String expected) {
```
wonder if we really need this method, can't we just expose MockTemplate?

martijnvg replied:
I did this to avoid using class names, but `new TestTemplateService.MockTemplate(...)` isn't that longer than `TestTemplateService.template(...)`. I'll remove it.

javanna commented:
I left a few minor comments but it's really close now. I think that we will have to see whether the templating logic needs to be moved to IngestDocument, I think it will be at some point, but for now it's ok as-is. Also, we will need to make adjustments and better document how other runtimes can use ingest-core, given that we use ScriptService and each runtime has to provide its own template service, but again for now it's ok I think.

martijnvg commented:
@javanna I've updated the PR based on your comments.

> I think that we will have to see whether the templating logic needs to be moved to IngestDocument

Agreed, it is trivial now the cut other processor settings over to use templates.

> Also, we will need to make adjustments and better document how other runtimes can use ingest-core, given that we use ScriptService and each runtime has to provide its own template service, but again for now it's ok I think.

Yes, we are likely going to makes changes, once the ingest framework is going to be ran in other runtimes. The main achieved goal here is that the templating in ingest isn't tied to ES.

javanna commented:
LGTM go @martijnvg go
